network/src/bounded_executor.rs

```diff
@@ -81,24 +112,107 @@ impl BoundedExecutor {
         F::Output: Send + 'static,
     {
         // Release the permit back to the semaphore when this task completes.
-        let f = f.map(move |ret| {
+        let f = Self::with_permit(f, spawn_permit);
+        self.executor.spawn(f)
+    }
+
+    // Returns a [`Future`] that complies with the `BoundedExecutor`. Once launched,
+    // will block if the executor is at capacity, until one of the other spawned
+    // futures completes.
+    async fn run_on_semaphore<F>(semaphore: Arc<Semaphore>, f: F) -> F::Output
+    where
+        F: Future + Send + 'static,
+        F::Output: Send + 'static,
+    {
+        let permit = Self::acquire_permit(semaphore.clone()).await;
+        Self::with_permit(f, permit).await
+    }
+
+    /// Spawn a [`Future`] on the `BoundedExecutor`. This function is async and
+    /// will block if the executor is at capacity until one of the other spawned
+    /// futures completes.
+    /// In case the future completes with an error, it retries, queueing attempts
+    /// on the semaphore according to the provided [`crate::RetryConfig`].
+    /// This function returns a [`JoinHandle`] that the caller can `.await` on for
+    /// the results of the overall set of attempts to execute the [`Future`].
+    pub async fn spawn_with_retries<F, Fut, T, E>(
+        &self,
+        retry_config: crate::RetryConfig,
+        mut f: F,
+    ) -> JoinHandle<Result<T, E>>
+    where
+        F: FnMut() -> Fut + Send + 'static,
+        Fut: Future<Output = Result<T, backoff::Error<E>>> + Send + 'static,
+        T: Send + 'static,
+        E: Send + 'static,
+    {
+        let retrier = || {
+            let semaphore = self.semaphore.clone();
+
+            let executor = move || {
+                let semaphore = semaphore.clone();
+                BoundedExecutor::run_on_semaphore(semaphore, f())
+            };
+
+            retry_config.retry(executor)
+        };
+        self.executor.spawn(retrier())
+    }
+
+    // Equips a future with a final step that drops the held semaphore permit
```

Comment:
Since we have the `.await` in the end, it equips and executes the future, right?

Reply:
This is an async function, so no execution here.